shared/domains/chelonia/chelonia.js

```diff
@@ -1642,50 +1646,110 @@ export default (sbp('sbp/selectors/register', {
+  // `data` is allowed to be falsy, in which case a fetch will occur first and
+  // the `onconflict` handler will be called.
   'chelonia/kv/set': async function (contractID: string, key: string, data: Object, {
+    ifMatch,
     innerSigningKeyId,
     encryptionKeyId,
     signingKeyId,
     maxAttempts,
     onconflict
   }: {
+    ifMatch?: string,
     innerSigningKeyId: ?string,
     encryptionKeyId: ?string,
     signingKeyId: string,
     maxAttempts: ?number,
-    onconflict: (contractID: string, key: string, data: Object) => Promise<Object>,
+    onconflict: ChelKvOnConflictCallback,
   }) {
     maxAttempts = maxAttempts ?? 3
+    const url = `${this.config.connectionURL}/kv/${encodeURIComponent(contractID)}/${encodeURIComponent(key)}`
     for (;;) {
-      const serializedData = outputEncryptedOrUnencryptedMessage.call(this, {
-        contractID,
-        innerSigningKeyId,
-        encryptionKeyId,
-        signingKeyId,
-        data,
-        meta: key
-      })
-      const response = await this.config.fetch(`${this.config.connectionURL}/kv/${encodeURIComponent(contractID)}/${encodeURIComponent(key)}`, {
-        headers: new Headers([[
-          'authorization', buildShelterAuthorizationHeader.call(this, contractID)
-        ]]),
-        method: 'POST',
-        body: JSON.stringify(serializedData),
-        signal: this.abortController.signal
-      })
+      let response: Response
+      if (data || typeof onconflict !== 'function') {
+        const serializedData = outputEncryptedOrUnencryptedMessage.call(this, {
+          contractID,
+          innerSigningKeyId,
+          encryptionKeyId,
+          signingKeyId,
+          data,
+          meta: key
+        })
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ], [
+            'if-match', ifMatch || '""'
+          ]
+          ]),
+          method: 'POST',
+          body: JSON.stringify(serializedData),
+          signal: this.abortController.signal
+        })
+      } else {
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ]]),
+          signal: this.abortController.signal
+        })
+      }
+      const resolveData = async () => {
+        let currentValue
+        if (response.ok || response.status === 409 || response.status === 412) {
+          const serializedData = await response.json()
+          currentValue = parseEncryptedOrUnencryptedMessage.call(this, {
+            contractID,
+            serializedData,
+            meta: key
+          })
+        } else if (response.status !== 404 && response.status !== 410) {
+          throw new ChelErrorUnexpectedHttpResponseCode('[kv/set] Invalid response code: ' + response.status)
+        }
```

Comment:
This whole section is a bit confusing to me as it's not clear to me what these response codes signify in terms of what was sent to the server.

E.g. why is 404 being paired with 410 here? And in what situation would 404 be returned?

Are 409 and 412 the same situation, if not, how are they different, and if they are, why do we use different codes?

Can you add a comment above these cases explaining all of this? Would make it a lot easier to read and reason about without having to jump to the server code.

Reply:
I'll add comments, but this isn't sending anything to the server though, it's parsing a response.

404 and 410 are paired because they're very similar, and represent there's no data (so new data can be set).

409 and 412 represent conflicts, although of different sorts. 409 is a height conflict while 412 is an if-match (CID / ETag) conflict.